### PR TITLE
ARROW-7836: [Rust] "allocate_aligned"/"reallocate" need to initialize memory to avoid UB

### PR DESCRIPTION
**Pull request by paddyhoran**

### CODE REVIEW

**Comment by paddyhoran** [comment]
The CI failures are unrelated.  @Marwes @jturner314 are you happy with this solution?

File: rust/arrow/src/memory.rs
Comment (jturner314) [inline rust/arrow/src/memory.rs]: Fwiw, using the [`add`](https://doc.rust-lang.org/std/primitive.pointer.html#method.add) method is equivalent and a bit more concise:
```suggestion
                new_ptr.add(old_size),
```

File: rust/arrow/src/memory.rs
Comment (jturner314) [inline rust/arrow/src/memory.rs]: There should be an additional check here in case `realloc` [returns a null pointer (indicating an error)](https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#method.realloc), since otherwise the [`offset` will be out-of-bounds](https://doc.rust-lang.org/std/primitive.pointer.html#safety-1). (Also, `std::ptr::write_bytes` [requires the pointer to be non-null](https://doc.rust-lang.org/std/ptr/fn.write_bytes.html#safety).)
```suggestion
        if !new_ptr.is_null() && new_size > old_size {
```

File: rust/arrow/src/memory.rs
Comment (jturner314) [inline rust/arrow/src/memory.rs]: Fwiw, using the [`.write_bytes()`](https://doc.rust-lang.org/std/primitive.pointer.html#method.write_bytes) method on the pointer would be a bit more concise.

**Review by jturner314 — COMMENTED** [review]
With the addition of a null pointer check (see the comment), this looks good to me. I think zeroing out all new memory, as in this PR, is the safest option. My understanding is that zeroing memory is very fast. At some later point, it may be worth revisiting this to see how much of a performance impact it has. (For example, it may be sufficient to just zero out the padding used for SIMD, instead of the entire allocation.) But, without a more thorough review of accesses into memory buffers, I'd prioritize safety over potential performance gains.

**Comment by paddyhoran** [comment]
I addressed all the comments.  @jturner314 thank you very much for all your time and help!

> I'd prioritize safety over potential performance gains.

Yea, definitely my thinking too.

**Comment by paddyhoran** [comment]
@andygrove, @nevi-me @sunchao can I get at least one approval from another committer to get this merged if you have a chance?

**Review by nevi-me — APPROVED** [review]

**Pull request closed without merging**
